File: v2/src/main/java/com/api/v2/customers/services/CustomerRegistrationServiceImpl.java
package com.api.v2.customers.services;

import com.api.v2.customers.domain.CustomerRepository;
import com.api.v2.customers.domain.exposed.Customer;
import com.api.v2.customers.responses.CustomerResponseDto;
import com.api.v2.people.exceptions.DuplicatedEmailException;
import com.api.v2.people.exceptions.DuplicatedSsnException;
import com.api.v2.people.requests.PersonRegistrationDto;
import com.api.v2.people.services.exposed.PersonRegistrationService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.net.URI;

@Service
@RequiredArgsConstructor
public class CustomerRegistrationServiceImpl implements CustomerRegistrationService {

    private final PersonRegistrationService personRegistrationService;
    private final CustomerRepository customerRepository;

    @Override
    public Mono<ResponseEntity<CustomerResponseDto>> register(@Valid PersonRegistrationDto registrationDto) {
        return validate(registrationDto)
                .then(personRegistrationService.register(registrationDto)
                .flatMap(person -> {
                    Customer customer = Customer.of(registrationDto.address(), person);
                    return customerRepository
                            .save(customer)
                            .flatMap(savedCustomer -> {
                                CustomerResponseDto responseDto = savedCustomer.toDto();
                                ResponseEntity<CustomerResponseDto> responseEntity = ResponseEntity
                                        .created(URI.create("/api/v2/customers"))
                                        .body(responseDto);
                                return Mono.just(responseEntity);
                            });
                }));
    }

    private Mono<ResponseEntity<CustomerResponseDto>> validate(PersonRegistrationDto registrationDto) {
        return customerRepository
                .findBySsn(registrationDto.ssn())
                .switchIfEmpty(Mono.error(DuplicatedSsnException::new))
                .flatMap(ignored -> Mono.empty())
                .then(customerRepository
                        .findByEmail(registrationDto.email())
                        .switchIfEmpty(Mono.error(DuplicatedEmailException::new))
                        .flatMap(ignored -> Mono.empty())
                );
    }

}

File: v2/src/main/kotlin/com/api/v2/customers/services/CustomerRegistrationServiceImpl.kt
package com.api.v2.customers.services

import com.api.v2.customers.Customer
import com.api.v2.customers.domain.CustomerCrudRepository
import com.api.v2.customers.CustomerResponseDTO
import com.api.v2.people.DuplicatedSINException
import com.api.v2.people.requests.PersonRegistrationDTO
import com.api.v2.people.PersonRegistrationService
import com.api.v2.customers.toDTO
import jakarta.validation.Valid
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import java.net.URI

@Service
class CustomerRegistrationServiceImpl(
    val repository: CustomerCrudRepository,
    val personRegistrationService: PersonRegistrationService
) : CustomerRegistrationService {

    override fun register(registrationDTO: @Valid PersonRegistrationDTO): ResponseEntity<CustomerResponseDTO> {
        validate(registrationDTO)
        val savedPerson = personRegistrationService.register(registrationDTO)
        val newCustomer = Customer.update(savedPerson)
        val savedCustomer = repository.save<Customer>(newCustomer)
        val dto = savedCustomer.toDTO()
        return ResponseEntity
            .created(URI.create("/api/v2/customers"))
            .contentType(MediaType.APPLICATION_JSON)
            .body(dto)
    }

    private fun validate(registrationDTO: PersonRegistrationDTO) {
        if (repository.findBySIN(registrationDTO.sin) != null) {
            throw DuplicatedSINException()
        }
        if (repository.findByEmail(registrationDTO.email) != null) {
            throw DuplicatedSINException()
        }
    }
}